Simultaneous library load causes persistent plugin load failure

cottsay opened the pull request:
I've been hitting this issue for about a year now, and I figured that it was serious enough that it would be caught and fixed quickly, but its still in the latest class_loader from groovy-devel.

The problem presents itself sporadically when loading nodelets with a message indicating that no factory function exists for the target class. When I trace the problem, it appears to happen whenever a library load is attempted while another library load is in progress. From what I can tell:

Plugin A enters https://github.com/ros/class_loader/blob/groovy-devel/src/class_loader_core.cpp#L432

Plugin B enters https://github.com/ros/class_loader/blob/groovy-devel/src/class_loader_core.cpp#L432 before plugin A reaches https://github.com/ros/class_loader/blob/groovy-devel/src/class_loader_core.cpp#L455

Plugin A then enters https://github.com/ros/class_loader/blob/groovy-devel/src/class_loader_core.cpp#L464 even though it is the initial load, and thus never registers any Meta Objects, which causes all future attempts to load the plugin to fail while this instance of the class_loader_private is in memory.

I created a static mutex in the same way the others are made in this file, and turned https://github.com/ros/class_loader/blob/groovy-devel/src/class_loader_core.cpp#L430-L457 into a critical section of code. This seems to prevent the scenario from occurring.

I think the easiest way to trick this into happening is to make the library load take a long time (ie slow hard disk), and try to load two plugins from different libraries at the same time. This has been quite difficult to reproduce. I think the reason it happens on our lab machines is that their ROS environments are NFS mounted, and so there is a little bit of network latency associated with file access, but I can't be sure.

Am I crazy? Is this a poor approach to solving this problem?

Thanks!

--scott

mirzashah commented:
@cottsay No you're not crazy! Thanks for bringing this up...I didn't realize it was a problem. Sounds like a mutex will fix the code...I'll take a look at it in the morning.

mirzashah commented:
@cottsay Sorry for the delay in fixing this, I'm a bit backed up. If you wouldn't mind, could you send me a pull request with your fix -- it's easier for me to just review it, run it, and then merge it. Otherwise it's gonna keep going on the back burner for me :/

cottsay committed:
Serialize initial library loads

class_loader_core uses static entities to create global objects within the class. Well it turns out that the loadLibrary function can be called multiple times simultaneously, and because of the static objects, we end up starting a new load before another finishes, causing a persistent failure when loading the first library.

This commit adds a mutex to ensure that this critical section of the loadLibrary function is not interrupted by another library load.

cottsay commented:
Okay, I attached my fix. The diff makes the end of the change look a little bit strange, but all I really did was indent the try/catch block and two lines from below, and make them a critical section.

Not sure how you feel about "arbitrarily" scoping the section, but it sure looks a lot cleaner than unlocking the mutex in each catch...

mirzashah commented:
Oh awesome, thanks so much...I was coincidentally going to be looking at this this week. I'm fine with the "arbitrary scope block" + the scoped lock...I would do the same thing.

dirk-thomas commented on src/class_loader_core.cpp:
```diff
@@ -427,33 +434,38 @@ void loadLibrary(const std::string& library_path, ClassLoader* loader)
-  }
-  catch(const Poco::NotFoundException& e)
   {
+    boost::recursive_mutex::scoped_lock libload_lock(getLibraryLoaderMutex());
```
Since `getLibraryLoaderMutex()` is only used in this one location why not use a local static variable?

```
static boost::recursive_mutex load_mutex;
boost::recursive_mutex::scoped_lock load_lock(mutex);
```

cottsay replied:
This means less changes for sure. I can't see any other places where we would need to use this mutex, so this seems like a good plan.

dirk-thomas replied:
Could you please update your pull request with that and afterwards squash your commits?

cottsay commented:
Well, I squashed the commits in https://github.com/cottsay/class_loader, but it appears that since the base commit of the pull request changed, this pull request didn't update to see the new code.

Not sure if there is a _special_ way to squash on Github, but that surely didn't work.

Suggestions? Shall I open a new pull?

tfoote commented:
You can use the git rebase to update the base commit of your branch.

cottsay commented:
The command I used to squash was `git rebase -i HEAD~2`, and it did exactly what it should have to my fork. The issue is that this pull request no longer appears from my fork, so changes I make to my fork won't appear here. I have no way to modify the pull request any longer. There seems to be a temporary tree on the main repo that is used to store the commit, and I can't even pull that tree down.

I must have mucked this one up somehow, and I'm sorry about that. Not quite a git master yet...

cottsay commented:
Moved to #16
